tests/cases/compiler/inKeywordNarrowingWithNoUncheckedIndexedAccess.ts

```diff
@@ -0,0 +1,37 @@
+// @strict: true
+// @noEmit: true
+// @noUncheckedIndexedAccess: true
+
+declare function invariant(condition: boolean): asserts condition;
+
+function f1(obj: Record<string, string>) {
+    invariant("test" in obj);
+    return obj.test;  // string
+}
+
+function f2(obj: Record<string, string>) {
+    if ("test" in obj) {
+        return obj.test;  // string
+    }
+    return "default";
+}
+
+function f3(obj: Record<string, string>) {
+    obj.test;  // string | undefined
+    if ("test" in obj) {
+        obj.test;  // string
+    }
+    else {
+        obj.test;  // undefined
+    }
+}
+
+function f4(obj: Record<string, string>) {
+    obj.test;  // string | undefined
+    if (obj.hasOwnProperty("test")) {

```

Comment:
wow, how does this one work if `hasOwnProperty`'s signature is "just":
> hasOwnProperty(v: PropertyKey): boolean;

what kind of sorcery is going on here? :D does this method have some special support in the compiler itself?

Reply:
> does this method have some special support in the compiler itself?

It does indeed.

Reply:
I’m going to assume that feature was implemented early on since nowadays feature requests to special-case built-in methods get rejected if the special case can’t be represented in the method signature.

Reply:
For reference, I found the code related to `hasOwnProperty` [here](https://github.dev/microsoft/TypeScript/blob/f6628a4573cd37c26912f78de3d08cd1dbf687a5/src/compiler/checker.ts#L26781)